jvb/src/main/kotlin/org/jitsi/videobridge/EndpointK.kt

```diff
@@ -71,24 +83,144 @@ import org.jitsi_modified.sctp4j.SctpDataCallback
+
+    /**
+     * The [SctpSocket] for this endpoint, if an SCTP connection was
+     * negotiated.
+     */
+    private var sctpSocket: Optional<SctpServerSocket> = Optional.empty()
+
+    /**
+     * The set of [ChannelShim]s associated with this endpoint. This
+     * allows us to expire the endpoint once all of its 'channels' have been
+     * removed. The set of channels shims allows to determine if endpoint
+     * can accept audio or video.
+     */
+    private val channelShims = mutableSetOf<ChannelShim>()
+
+    /**
+     * Whether this endpoint should accept audio packets. We set this according
+     * to whether the endpoint has an audio Colibri channel whose direction
+     * allows sending.
+     */
+    private var acceptAudio = false
+    /**
+     * Whether this endpoint should accept video packets. We set this according
+     * to whether the endpoint has a video Colibri channel whose direction
+     * allows sending.
+     */
+    private var acceptVideo = false
 
-    // TODO: this naming is to avoid conflicts with getTransceiver in Endpoint.  It will change back
-    // once Endpoint.java goes away
-    val _transceiver = Transceiver(
+    /**
+     * The queue we put outgoing SRTP packets onto so they can be sent
+     * out via the [IceTransport] on an IO thread.
+     */
+    private val outgoingSrtpPacketQueue = PacketInfoQueue(
+        "${javaClass.simpleName}-outgoing-packet-queue",
+        TaskPools.IO_POOL,
+        this::doSendSrtp,
+        TransportConfig.queueSize
+    ).apply {
+        setErrorHandler(queueErrorCounter)
+    }
+
+    private val bitrateController = BitrateController(
+        object : BitrateController.EventHandler {
+            override fun allocationChanged(allocation: BandwidthAllocation) {
+                // Intentional no-op
+            }
+
+            override fun forwardedEndpointsChanged(forwardedEndpoints: Set<String>) =
+                sendForwardedEndpointsMessage(forwardedEndpoints)
+
+            override fun effectiveVideoConstraintsChanged(
+                oldEffectiveConstraints: Map<String, VideoConstraints>,
+                newEffectiveConstraints: Map<String, VideoConstraints>
+            ) = this@EndpointK.effectiveVideoConstraintsChanged(oldEffectiveConstraints, newEffectiveConstraints)
+
+            override fun keyframeNeeded(endpointId: String?, ssrc: Long) =
+                conference.requestKeyframe(endpointId, ssrc)
+        },
+        Supplier { getOrderedEndpoints() },
+        diagnosticContext,
+        logger
+    )
+
+    /**
+     * The instance which manages the Colibri messaging (over a data channel
+     * or web sockets).
+     */
+    private val _messageTransport = EndpointMessageTransport(
+        this,
+        Supplier { conference.videobridge.statistics },
+        conference,
+        logger
+    )
+
+    override fun getMessageTransport(): EndpointMessageTransport = _messageTransport
```

Comment:
Hm, you can't override `getMessageTransport()` with a `val`?

Reply:
No, unfortunately that doesn't seem to work.

Reply:
Annoying, okay.